### PR TITLE
fix: Allow numbers in steps.args.params.value

### LINKED ISSUE #2413: New Argo releases fail on workflows with parameters that are not strings
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [ ] I've included the logs.
 
**What happened**:
Create a simple workflow where input paramters contain, say, a number and submit it either via the CLI or via the Golang API. Argo will reject it saying that steps can't be numbers.

**What you expected to happen**:
I expect to see the workflow get compiled and run to completion.

**How to reproduce it (as minimally and precisely as possible)**:
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: hello-world-parameters-
spec:
  entrypoint: debug
  templates:
  - name: debug
    steps:
    - - arguments:
          parameters:
          - name: p1
            value: '1'
          - name: p2
            value: 1
        name: hello
  - container:
      command:
      - echo "{{inputs.parameters.p1}}, {{inputs.parameters.p1}}"
      image: docker/whalesay
      parameters:
      - name: p1
      - name: p2
```

**Anything else we need to know?**:
We have tested this on Argo 2.5.2 and Argo 2.6.0

**Environment**:
- Argo version:
```
argo: v2.6.0
  BuildDate: 2020-02-28T23:32:45Z
  GitCommit: 5d3bdd56607eea962183a9e45009e3d08fafdf9b
  GitTreeState: clean
  GitTag: v2.6.0
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```

- Kubernetes version :
```
clientVersion:
  buildDate: "2019-10-15T23:41:55Z"
  compiler: gc
  gitCommit: c97fe5036ef3df2967d086711e6c0c405941e14b
  gitTreeState: clean
  gitVersion: v1.16.2
  goVersion: go1.12.10
  major: "1"
  minor: "16"
  platform: darwin/amd64
serverVersion:
  buildDate: "2019-11-14T04:58:54Z"
  compiler: gc
  gitCommit: 2640ac46d96791a135961425127d9c2f7e184924
  gitTreeState: clean
  gitVersion: v1.15.5
  goVersion: go1.12.10
  major: "1"
  minor: "15"
  platform: linux/amd64
```

**Other debugging information (if applicable)**:
- workflow result:
Not Applicable (N/A)

- executor logs:
Not Applicable (N/A)

- workflow-controller logs:
Not Applicable (N/A)

**Logs**
The workflow never got submitted but here is what Argo tells us:
```
2020/03/11 12:01:48 Failed to parse workflow: error unmarshaling JSON: while decoding JSON: json: cannot unmarshal number into Go struct field Template.spec.templates.steps of type string
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by simster7**
Fixes #2413 

In versions <v2.5.0 we allowed the step.arguments.parameters.value field to be a number as well as a string. In version >=v2.5.0 a custom unmarshaller was implemented that enforced stricter type checking, but we still want to add backwards-compatibility to workflows. Here we manually attempt to convert a number to a string field.

This current solution is hacky, and am open to suggestions in improving it.

### CODE REVIEW

**Comment by sarabala1979** [comment]
I would recommend adding `UnMarshalJSON` function for the `parameter` to convert values to a string.

workflow_type.go
```
func (p *Parameter) UnmarshalJSON(value []byte) error {
	var candidate map[string]interface{}
	err := json.Unmarshal(value, &candidate)
	if err != nil {
		return err
	}
	p.Name =  fmt.Sprint(candidate["name"])
	itemValue := fmt.Sprint(candidate["value"])
	p.Value = &itemValue
	return nil
}

```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -540,6 +539,39 @@ type Parameter struct {
...
+		if err != nil {
+			return err
+		}
+		err = json.Unmarshal(strVal, &p.ValueFrom)
+		if err != nil {
+			return err
+		}
+	}
+	if val, ok := candidate["globalName"]; ok {
+		p.Name = fmt.Sprint(val)
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: wrong variable. 'p.globalName`

Comment (simster7) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Nice catch

**Comment by simster7** [comment]
Back-ported to 2.6

**Pull request merged**